// @wordpress/block-editor/build/components/warning/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _reactNative = require("react-native");
var _components = require("@wordpress/components");
var _compose = require("@wordpress/compose");
var _blocks = require("@wordpress/blocks");
var _style = _interopRequireDefault(require("./style.scss"));
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function Warning({
  actions,
  title,
  message,
  icon,
  iconClass,
  preferredColorScheme,
  getStylesFromColorScheme,
  containerStyle: extraContainerStyle,
  titleStyle: extraTitleStyle,
  messageStyle: extraMessageStyle,
  ...viewProps
}) {
  icon = icon && (0, _blocks.normalizeIconObject)(icon);
  const internalIconClass = 'warning-icon' + '-' + preferredColorScheme;
  const containerStyle = [getStylesFromColorScheme(_style.default.container, _style.default.containerDark), extraContainerStyle];
  const titleStyle = [getStylesFromColorScheme(_style.default.title, _style.default.titleDark), extraTitleStyle];
  const messageStyle = [getStylesFromColorScheme(_style.default.message, _style.default.messageDark), extraMessageStyle];
  return (0, _react.createElement)(_reactNative.View, {
    style: containerStyle,
    ...viewProps
  }, icon && (0, _react.createElement)(_reactNative.View, {
    style: _style.default.icon
  }, (0, _react.createElement)(_components.Icon, {
    className: iconClass || internalIconClass,
    icon: icon && icon.src ? icon.src : icon
  })), title && (0, _react.createElement)(_reactNative.Text, {
    style: titleStyle
  }, title), message && (0, _react.createElement)(_reactNative.Text, {
    style: messageStyle
  }, message), actions);
}
var _default = exports.default = (0, _compose.withPreferredColorScheme)(Warning);
