// @wordpress/block-editor/build-module/components/warning/index.native.js
import { createElement } from "react";
/**
 * External dependencies
 */
import { View, Text } from 'react-native';

/**
 * WordPress dependencies
 */
import { Icon } from '@wordpress/components';
import { withPreferredColorScheme } from '@wordpress/compose';
import { normalizeIconObject } from '@wordpress/blocks';

/**
 * Internal dependencies
 */
import styles from './style.scss';
function Warning({
  actions,
  title,
  message,
  icon,
  iconClass,
  preferredColorScheme,
  getStylesFromColorScheme,
  containerStyle: extraContainerStyle,
  titleStyle: extraTitleStyle,
  messageStyle: extraMessageStyle,
  ...viewProps
}) {
  icon = icon && normalizeIconObject(icon);
  const internalIconClass = 'warning-icon' + '-' + preferredColorScheme;
  const containerStyle = [getStylesFromColorScheme(styles.container, styles.containerDark), extraContainerStyle];
  const titleStyle = [getStylesFromColorScheme(styles.title, styles.titleDark), extraTitleStyle];
  const messageStyle = [getStylesFromColorScheme(styles.message, styles.messageDark), extraMessageStyle];
  return createElement(View, {
    style: containerStyle,
    ...viewProps
  }, icon && createElement(View, {
    style: styles.icon
  }, createElement(Icon, {
    className: iconClass || internalIconClass,
    icon: icon && icon.src ? icon.src : icon
  })), title && createElement(Text, {
    style: titleStyle
  }, title), message && createElement(Text, {
    style: messageStyle
  }, message), actions);
}
export default withPreferredColorScheme(Warning);
